src/routes/(blog-article)/containerizing-rust-applications-best-practices/+page.md

```diff
@@ -0,0 +1,1049 @@
+
+4. **Accidental Damage**: Even if an attacker does not compromise the container, running as root increases the risk of accidental damage by the container's own applications or administrators. For example, a poorly crafted command could delete critical files or disrupt important processes.
+
+There are some ways to avoid running the container as `root`. We will see all of them.
+
+### Use the `USER` instruction
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+FROM rust:latest
+
+WORKDIR /app
+
+RUN cargo init
+RUN cargo build --release
+
+USER www-data
+
+CMD ["./target/release/app"]
+```
+
+</CodeBlock>
+
+You can add the `USER` instruction before the last command. In that example we
+know that the base image already contains the user `www-data`.
+
+<CodeBlock lang="console">
+
+```console
+docker run --rm -it docker-rust-app-running-with-root cat /etc/passwd | grep www-data
+www-data:x:33:33:www-data:/var/www:/usr/sbin/nologin
+```
+
+</CodeBlock>
+
+But you can also create a specific user for your application:
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+FROM rust:latest
+
+WORKDIR /app
+
+RUN cargo init
+RUN cargo build --release
+
+RUN groupadd -r appuser
+RUN useradd -r -u 1001 -g appuser appuser
+
+USER appuser
+
+CMD ["./target/release/app"]
+```
+
+</CodeBlock>
+
+Using the `USER` instruction is considered a good practice because by default the container
+will not run as root. In that example, docker will run the container as the user with the ID `1001`.
+
+<Callout type="info">
+
+The Linux Standard Base Core Specification defines three different ranges for
+user IDs, the value from 0 to 99, 100 to 999, and 1000 and above. The first two
+ranges are reserved for system users. All the regular users start from 1000.
+It's also important to note that the user ID 0 is reserved for the root user and
+Linux kernel uses the user id and group id to handle the permissions. The host
+machine and the docker container might have the same user ID but different usernames.
+
+</Callout>
+
+If you run that container, `appuser` will have the same permissions as the user `1001`
+in the host system, whatever it's the name of that user in the host machine.
+
+### Use The `docker run --user` Argument
+
+You can also overwrite the user running the container with the argument `--user`:
+
+<CodeBlock lang="console">
+
+```console
+docker run --user www-data --rm -it docker-rust-app-running-with-root whoami
+www-data
+```
+
+</CodeBlock>
+
+In this example, even though the image is execute as root by default, it will be
+executed as the `www-data` user.
+
+Notice you can even use a non-existing user in both the host and the docker image.
+
+<CodeBlock lang="console">
+
+```console
+$ docker run --user 1001 --rm -it docker-rust-app-running-with-root bash
+I have no name!@895b0f6a3dbb:/app$ 
+```
+
+</CodeBlock>
+
+All this solutions work but they all have a drawback: you need to know the user
```

Comment:
If you use plural it should read "these" instead of "this"
otherwise take away the plural "s" of solutions.